designs/2021-break-on-parsing-errors/README.md

```diff
@@ -0,0 +1,185 @@
+- Repo: eslint/eslint
+- Start Date: 2021/1/12
+- RFC PR: https://github.com/eslint/rfcs/pull/75
+- Authors: [A-Katopodis](https://github.com/A-Katopodis)
+
+# (Break on parsing errors)
+
+## Summary
+
+The suggested change is for ESLint tp support an argument that will make the CLI exit with code 2 if any fatal errors are reported. The option will be an opt-in boolean argument called `--fatal-parse-error`.
+
+## Motivation
+
+We met with a couple of cases where we assumed a succesfull run of ESLint in CI enviroments. When ESLint wouldn't be able to read the tsconfig.json or had a wrongly configured source type the error would be a `fatal` one but the exit code would be `1`. There was no distiction between a run with no fatal errors and with fatal errors.
+
+According to the eslint docs about exit codes:
+
+- 0: Linting was successful and there are no linting errors. If the --max-warnings flag is set to n, the number of linting warnings is at most n.
+- 1: Linting was successful and there is at least one linting error, or there are more linting warnings than allowed by the --max-warnings option.
+- 2: Linting was unsuccessful due to a configuration problem or an internal error.
+
+Being able to exit with code `2` instead of `1` it will allow for some CI pipelines to better understand the results and react differently if ESLint reports any `fatal` errors. 
+
+Altough it is possible to filter out the results it will not allow the user 
+
+## Detailed Design
+
+Design Summary:
+
+1. Add a `fatal-parse-error` option.
+2. Gather and report the fatal messages from `cli-engine`.
+3. Return exit code `2` in the case of any fatal errors on `cli`
```

Comment:
Should this option apply to all `fatal: true` linting messages?

Aside from [parsing errors](https://github.com/eslint/eslint/blob/b4e2af5db1c29343ffec2cd104b04bf39b77ee56/lib/linter/linter.js#L690), there is at least one other case where ESLint produces linting messages with `fatal: true` - when a configuration comment [cannot be JSON-parsed](https://github.com/eslint/eslint/blob/b4e2af5db1c29343ffec2cd104b04bf39b77ee56/lib/linter/config-comment-parser.js#L104) (though this is also some sort of a parsing error).

Reply:
I think it should be to all `fatal: true` messages.

Any type of error (coming from invalid JSON, invalid `sourceType` or a parser error) that does not allow the file to be linted and the rules to run should be reported.

Reply:
ESLint will still lint a file with an invalid configuration comment. It can be said that the file wasn't successfully linted because the configuration comment wasn't applied, but there will be some inconsistencies.

For example: 

1.

```js
/* eslint no-undef: ["error" */ // <-- missing ']', invalid JSON 

/* eslint no-void: ["error"] */

void a;
```

2.

```js
/* eslint no-undef: ["error", "foo"] */ // <-- invalid option

/* eslint no-void: ["error"] */

void a;

```

3.

```js
/* eslint no-undeff: ["error"] */ // <-- typo, this rule doesn't exist

/* eslint no-void: ["error"] */

void a;

```

In all three examples, ESLint lints the file and reports a `no-void` linting error. It also reports a linting error for the invalid configuration comment on the first line in all three examples, but only in the first example the error has `fatal: true`.

Reply:
The behavior in those three cases makes sense to me, in terms of severity.